unix/c_editline.c

```diff
@@ -22,17 +22,73 @@ void go_libedit_set_string_array(char **ar, int p, char *s) {
     ar[p] = s;
 }
 
-int go_libedit_get_clientdata(EditLine *el) {
-    void* p;
-    el_get(el, EL_CLIENTDATA, &p);
-    return (int)(intptr_t)p;
+struct clientdata {
+    int id;
+    int bracketed_paste;
+};
+
+struct clientdata* go_libedit_get_clientdata(EditLine *el) {
+    struct clientdata* cd;
+    el_get(el, EL_CLIENTDATA, &cd);
+    return cd;
 }
 
-void go_libedit_set_clientdata(EditLine *el, int v) {
-    void *p = (void*)(intptr_t) v;
-    el_set(el, EL_CLIENTDATA, p);
+int go_libedit_set_clientdata(EditLine *el, struct clientdata cd) {
+    struct clientdata* pcd = malloc(sizeof(struct clientdata));
+    if (pcd == NULL) {
+	return -1;
+    }
+    *pcd = cd;
+    return el_set(el, EL_CLIENTDATA, pcd);
+}
+
+/************* terminals ************/
+
+const char* term_bp_off = "\033[?2004l";
+const char* term_bp_on = "\033[?2004h";
+const char* term_bp_start = "\033[200~";
+const char* term_bp_end = "\033[201~";
+
+int go_libedit_term_supports_bracketed_paste(EditLine* el) {
+    // terminfo does not track support for bracketed paste, so we manually
+    // whitelist known-good TERM values:
+    //
+    //     - "screen" matches screen and tmux.
+    //     - "xterm*" matches Terminal.app, iTerm2, GNOME terminal, and probably
+    //       others.
+    //
+    // Modern versions of all the programs listed above support bracketed paste;
+    // older versions harmlessly ignore the control codes.
+    char* term = getenv("TERM");
+    return strcmp(term, "screen") == 0 || strncmp(term, "xterm", 5) == 0;
 }
 
+static unsigned char _el_bracketed_paste(EditLine *el, int ch) {
+    int endsz = strlen(term_bp_end);
+    char* buf = NULL;
+    size_t bufsz = 0;
+    size_t i = 0;
+    // Read characters until the buffer ends with term_bp_end.
+    while (i < endsz || strncmp(&buf[i-endsz], term_bp_end, endsz) != 0) {
+	if (i >= bufsz) {
+	    bufsz = (bufsz == 0 ? 1024 : bufsz * 2);
+	    if ((buf = realloc(buf, bufsz)) == NULL)
```

Comment:
Where is `buf` freed?

1) If `insertstr` does it, add a comment. Also check if it gets freed in case of error.
2) if the realloc fails, add a `free`.

Reply:
Nope, `insertstr` doesn't do it. I was sloppy. Fixed.